chain/client/src/chunk_validation.rs

```diff
@@ -423,10 +435,31 @@ fn apply_result_to_chunk_extra(
 impl Client {
     /// Responds to a network request to verify a `ChunkStateWitness`, which is
     /// sent by chunk producers after they produce a chunk.
-    pub fn process_chunk_state_witness(&mut self, witness: ChunkStateWitness) -> Result<(), Error> {
+    pub fn process_chunk_state_witness(
+        &mut self,
+        witness: ChunkStateWitness,
+        peer_id: PeerId,
+    ) -> Result<(), Error> {
         // TODO(#10265): If the previous block does not exist, we should
         // queue this (similar to orphans) to retry later.
-        self.chunk_validator.start_validating_chunk(witness, self.chain.chain_store())
+        match self.chunk_validator.start_validating_chunk(
+            witness,
+            self.chain.chain_store(),
+            peer_id.clone(),
+        ) {
+            Err(err) => {
+                if !matches!(err, Error::NotAValidator) {
```

Comment:
Hmm, what are we exactly banning for here? The error NotAValidator is returned when... the *recipient* is not a validator, right? That seems less of a problem. Should we instead match for signature verification failure, which means whoever *sends* the message is not authorized to send it?

Reply:
That was actually weird to filter out NotAValidator. And it doesn't make sense to me to filter any errors, because invalid prev_block_hash will lead to DBError, and we still would need to ban peer. So I ban peer in case of any error now.

Reply:
Hmmm.... Are you sure we should ban peer in case of a DBError? If the prev block hash doesn't exist doesn't that just mean this node is behind?

Either way, I'm rather uncomfortable with banning peers aggressively, it can stall the network all by itself just because we forgot some corner case. I'd rather do this conservatively and maybe reactively. Doesn't validating the signature early already solve most of the issues? We won't proceed with any more work other than validating the signature. Sure, someone can spam us but that mostly is a cost for them - generally in spam protection we don't need to deal with excessive requests, but rather excessive responses.

Reply:
@robin-near Makes sense. Still, I want to have some punishment condition against like passing a super long vector of implicit transitions or useless partial state which can potentially lead to a lot of work being done. How about doing that in case of `Error::InvalidChunkStateWitness`?

If it is not restricted at all, a node can send invalid state witnesses to specific peer just to do some heavy computation, and still get enough endorsements from all others, so I'm not comfortable with such messages just floating around.